server/dex/dex.go

```diff
@@ -0,0 +1,347 @@
+	for i, assetConf := range cfg.Assets {
+		symbol := strings.ToLower(assetConf.Symbol)
+		ID := assetIDs[i]
+
+		// Create a new asset backend. An asset driver with a name matching the
+		// asset symbol must be available.
+		log.Infof("Starting asset backend %q...", symbol)
+		logger := cfg.LogBackend.SubLogger("ASSET", symbol)
+		be, err := asset.Setup(symbol, assetConf.ConfigPath, logger, cfg.Network)
+		if err != nil {
+			abort()
+			return nil, fmt.Errorf("failed to setup asset %q: %v", symbol, err)
+		}
+
+		if symbol == "dcr" {
+			var ok bool
+			dcrBackend, ok = be.(*dcrasset.Backend)
+			if !ok {
+				abort()
+				return nil, fmt.Errorf("dcr backend is invalid")
+			}
+		}
+
+		startSubSys(fmt.Sprintf("Asset[%s]", symbol), be)
+
+		ba := &asset.BackedAsset{
+			Asset: dex.Asset{
+				ID:       ID,
+				Symbol:   symbol,
+				LotSize:  assetConf.LotSize,
+				RateStep: assetConf.RateStep,
+				FeeRate:  assetConf.FeeRate,
+				SwapConf: assetConf.SwapConf,
+				FundConf: assetConf.FundConf,
+			},
+			Backend: be,
+		}
+
+		backedAssets[ID] = ba
+		lockableAssets[ID] = &swap.LockableAsset{
+			BackedAsset: ba,
+			CoinLocker:  dexCoinLocker.AssetLocker(ID).Swap(),
+		}
+	}
+
+	// Create DEXArchivist with the pg DB driver.
+	pgCfg := &pg.Config{
+		Host:          cfg.DBConf.Host,
+		Port:          strconv.Itoa(int(cfg.DBConf.Port)),
+		User:          cfg.DBConf.User,
+		Pass:          cfg.DBConf.Pass,
+		DBName:        cfg.DBConf.DBName,
+		HidePGConfig:  false,
+		QueryTimeout:  20 * time.Minute,
+		MarketCfg:     cfg.Markets,
+		CheckedStores: true,
+		Net:           cfg.Network,
+		FeeKey:        cfg.RegFeeXPub,
+	}
+	storage, err := db.Open(ctx, pgCfg)
+	if err != nil {
+		abort()
+		return nil, fmt.Errorf("db.Open: %v", err)
+	}
+
+	authCfg := auth.Config{
+		Storage:         storage,
+		Signer:          cfg.DEXPrivKey,
+		RegistrationFee: cfg.RegFeeAmount,
+		FeeConfs:        cfg.RegFeeConfirms,
+		FeeChecker:      dcrBackend.UnspentCoinDetails,
+	}
+
+	authMgr := auth.NewAuthManager(&authCfg)
+
+	// Create the swapper.
+	swapperCfg := &swap.Config{
+		Assets:           lockableAssets,
+		Storage:          storage,
+		AuthManager:      authMgr,
+		BroadcastTimeout: cfg.BroadcastTimeout,
+	}
+
+	swapper := swap.NewSwapper(swapperCfg)
+	startSubSys("Swapper", swapper)
+
+	// Markets
+	markets := make(map[string]*market.Market, len(cfg.Markets))
+	for _, mktInf := range cfg.Markets {
+		baseCoinLocker := dexCoinLocker.AssetLocker(mktInf.Base).Book()
+		quoteCoinLocker := dexCoinLocker.AssetLocker(mktInf.Quote).Book()
+		mkt, err := market.NewMarket(mktInf, storage, swapper, authMgr, baseCoinLocker, quoteCoinLocker)
+		if err != nil {
+			abort()
+			return nil, fmt.Errorf("NewMarket failed: %v", err)
+		}
+		markets[mktInf.Name] = mkt
+	}
+
+	now := encode.UnixMilli(time.Now())
+	for name, mkt := range markets {
+		startEpochIdx := 1 + now/int64(mkt.EpochDuration())
+		mkt.SetStartEpochIdx(startEpochIdx)
```

Comment:
I just don't know how an operator would reasonable set this, except for perhaps allowing the operator to specify an earliest absolute time, from which the different start epochs could be calculated for each market.  Yeah, punting I guess, but I'm not sure if it's needed as a config item.

Reply:
Find to punt. 

That's pretty much what I had in mind, except I was thinking of specifying a delay duration instead of absolute time, that way a production server could set it and forget it, e.g. "markets re-open 10 minutes after the server starts".